File: BT_RightSide_View.py
# Leetcode 199. Binary Tree Right Side View

# Time Complexity :  O(n) where n is the number of the nodes

# Space Complexity :
# BFS:: O(n/2) == O(n) where n is the number of the nodes that can be in the q at worst
# DFS :: O(h) where h is the height of the tree

# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Approach: DFS :: at every node if the current level matches the size of the result array, add the node
# to result. Increment the level by 1 and call dfs recursively on left and right node.
# BFS:: add the root to que, at every level append the last node to the result. If the left or 
# right child are present add them to the q. return result after q is empty 

# Your code here along with comments explaining your approach

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# DFS approach :: Traversing the tree recursively starting with the right sub tree first at every level
class Solution:
    result = None
    def rightSideView(self, root: TreeNode) -> List[int]:
        self.result = []
        if not root:
            return self.result
        self.dfs(root,0)
        return self.result
    
    def dfs(self,node,level):
        if not node:
            return
        # Updating rightSide view when the current level matches the size of the view array
        if level == len(self.result):
            self.result.append(node.val)
        # Traversing next level starting with right subtree    
        self.dfs(node.right,level+1)
        self.dfs(node.left,level+1)


# BFS approach
from collections import deque
class Solution:
    result = None
    def rightSideView(self, root: TreeNode) -> List[int]:
        q = deque()
        q.append(root)
        result = []
        while q:
            size = len(q)
            # At every level append the last node to the result
            for i in range(len(q)):
                node = q.popleft()
                if i == size-1:
                    result.append(node.val)
                # If the left or right child are present add them to the q
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
        # return result after q is empty 
        return result

File: bt_right_side_view.cpp
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
// Your code here along with comments explaining your approach
//Recursive solution
// Time Complexity : O(n)
// Space Complexity : O(h) in worst case then O(n).
//DFS solution
//Recursive Part-1: left root call left root first and then replace
//1. Call recursion on the left node of the root till null condition is met and save the result at each level
//2. Check whether the right root is present and then replace
//3. Call recursion on the right node of the root and repeat
//Recursive Part-2: right root call right root first and then replacement
//1. Call recursion on the right node of the root till null condition is met and save the result at each level
//2. Call recursion on the left node of the root and add the level if not present
//BFS
// Time Complexity : O(n)
// Space Complexity :O(n/2).
//1.Create a queue and add elements level by level
//2.Save the last element of the level in the result and return. Inorder to do that size variable is required.
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> result;
    vector<int> rightSideView(TreeNode* root) {
        //edge case
        if(root==NULL){
            return result;
        }
        int level = 0;
       // get_right_elements_dfs_right_entry_first(root, 0);
       // get_right_elements_dfs_left_entry_first (root, 0);
        
        //iterative
        // stack<TreeNode *> tree_stack;
        // stack<int> level_stack;
        // bool first_flag = false;
        // while(root !=NULL || !tree_stack.empty()){
        //     if(tree_stack.empty() && result.size()!=0) level++;
        //     while(root!=NULL){
        //         tree_stack.push(root);
        //         if(result.size() == level){
        //           result.push_back(root->val);
        //           level_stack.push(level);
        //         }
        //         cout<<level<<endl;
        //         level++;
        //         root = root->right;
        //     }
        //     root = tree_stack.top();
        //     tree_stack.pop();
        //     level=level_stack.top();
        //     level_stack.pop();
        //     root = root->left;
        // }
        
        //BFS solution
        
        queue<TreeNode*> myqueue; 
        myqueue.push(root);
        //result.push_back(root->val);
       // TreeNode * temp= new TreeNode();
        while(!myqueue.empty()){
           // TreeNode * temp= ;
            result.push_back(myqueue.back()->val);
            int size = myqueue.size();
            while(size>0){
                TreeNode * temp = myqueue.front();
                myqueue.pop();
                size--;
                if(temp->left != NULL) myqueue.push(temp->left);
                if(temp->right != NULL) myqueue.push(temp->right);
            }
            
        }
        
        return result;
    }

private:
    void get_right_elements_dfs_right_entry_first (TreeNode* root, int level){
        //returning condition
        if(root==NULL){
            return;
        }
        if(result.size()==level){
            result.push_back(root->val);
        }
        get_right_elements_dfs_right_entry_first(root->right, level+1);
        // internal pop operation
        get_right_elements_dfs_right_entry_first(root->left, level+1);
    }
    
    void get_right_elements_dfs_left_entry_first (TreeNode* root, int level){
        //returning condition
        if(root==NULL){
            return;
        }
        if(result.size()==level){
            result.push_back(root->val);
        }else{
            result[level] = root->val;
        }
        get_right_elements_dfs_left_entry_first(root->left, level+1);
        // internal pop operation
        get_right_elements_dfs_left_entry_first(root->right, level+1);
    }
};
